### PR TITLE
Copy-engine support

### PR DESCRIPTION
**Pull request by dsidler**

### CODE REVIEW

File: tests/unittests/test_copy_engine_ops.py [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+    shmem.barrier()
+    del shmem
+
+
+# ============================================================================
+# Edge Cases and Error Conditions
+# ============================================================================
+
+
+@pytest.mark.skip(reason="Zero-size transfers may cause SDMA queue issues - needs investigation")
```
Comment (dsidler) [inline tests/unittests/test_copy_engine_ops.py]: if passed all the way to sdma 0 will be interpreted as -1

File: test.py
Comment (dsidler) [inline test.py:1]: delete

File: iris/host/memory/symmetric_heap.py [hunk 2]
```diff
@@ -327,6 +331,36 @@ def get_heap_bases(self) -> torch.Tensor:
         self._ensure_peer_refresh_healthy()
         return self.heap_bases
 
+    def translate(self, ptr: int, from_rank: int, to_rank: int) -> int:
+        """
+        Translate a pointer address from one rank's address space to another.
+
+        This is useful for host-side SDMA operations where you need to convert
```
Comment (dsidler) [inline iris/host/memory/symmetric_heap.py:338]: nit

File: iris/mem/triton/ops.py [hunk 3]
```diff
@@ -450,7 +688,44 @@ def atomic_cas(pointer, cmp, val, from_rank, to_rank, heap_bases, sem=None, scop
...
+        packet_offset_bytes = base + offset
+
+        # Place command packet
+        sdma_utils.place_atomic_cas_packet(queue_ptr_u32, packet_offset_bytes, dst_ptr_val, cmp, val)
+
+        # Submit command
+        pending_wptr = base + offset + command_in_bytes
+        sdma_utils.submit(write_ptr, doorbell_ptr, committed_write_ptr, base, pending_wptr)
+
+        return cmp
```
Comment (dsidler) [inline iris/mem/triton/ops.py:721]: remove

File: iris/mem/triton/ops.py [hunk 4]
```diff
@@ -683,3 +958,40 @@ def atomic_max(
...
+        >>>         copy_engine_ctx=copy_engine_ctx, USE_COPY_ENGINE=True, CONTIGUOUS_COPY=True
+        >>>     )
+        >>>     # Wait for completion
+        >>>     iris.quiet(copy_engine_ctx, 1)
+    """
+    # Extract queue pointers from context
+    # Context layout: [queue_buf, rptr, wptr, doorbell, cached_wptr, committed_wptr]
+    handle = copy_engine_ctx + (sdma_ep.QUEUE_DEVICE_CTX_SIZE * to_rank)
+    read_ptr = tl.load(handle + 1).to(tl.pointer_type(tl.uint64))
+    committed_wptr = tl.load(handle + 5).to(tl.pointer_type(tl.uint64))
```
Comment (dsidler) [inline iris/mem/triton/ops.py:983]: why not write_ptr?

File: tests/examples/test_message_passing.py [hunk 5]
```diff
@@ -63,9 +67,18 @@ def run_message_passing_kernels(module, args):
         # Allocate flags on the symmetric heap
         flags = shmem.zeros((num_blocks,), device="cuda", dtype=torch.int32)
 
+        copy_engine_ctx = shmem.get_copy_engine_ctx()
+
+        producer_fn = getattr(module.producer_kernel, "fn", None)
+        producer_params = (
+            producer_fn.__code__.co_varnames if producer_fn and hasattr(producer_fn, "__code__") else tuple()
+        )
+        needs_copy_engine_arg = any(param in producer_params for param in ("copy_engine_handle_ptr", "copy_engine_ctx"))
```
Comment (dsidler) [inline tests/examples/test_message_passing.py:76]: make consistent

File: tests/examples/test_message_passing.py [hunk 6]
```diff
@@ -107,6 +125,8 @@ def run_message_passing_kernels(module, args):
             import gc
 
             gc.collect()
+            # Clear CUDA cache to free GPU memory between tests
+            torch.cuda.empty_cache()
```
Comment (dsidler) [inline tests/examples/test_message_passing.py:129]: nit

File: tests/unittests/test_copy_engine_ops.py [hunk 7]
```diff
@@ -0,0 +1,758 @@
...
+
+    shmem.barrier()
+
+    # Each rank should have received the other's data
+    num_blocks = triton.cdiv(num_elements, 128)
+    assert completion_flag.item() == num_blocks, f"Expected {num_blocks} signals, got {completion_flag.item()}"
+    expected = _make_expected(num_elements, torch.float32, dst.device) * (remote_rank + 1)
+    assert torch.allclose(dst, expected)
+
+    shmem.barrier()
```
Comment (dsidler) [inline tests/unittests/test_copy_engine_ops.py:756]: rename to ctx?

**Comment by dsidler** [comment]
@mawad-amd not sure how useful the changes and copy egine path for examples 07 and 10 are. I might revert those and only keep the changes for message-passing example

**Comment by mawad-amd** [comment]
Thanks, David! I think we will want to keep the benchmark you used for your results. The message passing one would be a great introductory example.

Will review this PR soon.

File: iris/device/sdma_utils.py [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+and submission operations.
+"""
+
+import triton
+import triton.language as tl
+from xio import sdma_ep
+
+
+@triton.jit
+def wait_cnt():
```
Comment (mawad-amd) [inline iris/device/sdma_utils.py:18]: Can we move this to iris/mem/utils.py? Can we also guard for cdna3/4 and throw/return something for anything else?
Something like https://github.com/ROCm/iris/blob/main/iris/mem/utils.py#L69

File: iris/host/iris.py [hunk 9]
```diff
@@ -918,6 +950,306 @@ def get_heap_bases(self):
...
+            if dst_rank is None:
+                raise ValueError("dst_rank must be provided when translate=True")
+            ptr = self.heap.translate(ptr, self.get_rank(), dst_rank)
+
+        if ptr == 0:
+            return 0, 0
+        return ptr, bits
+
+    def put(
+        self,
```
Comment (mawad-amd) [inline iris/host/iris.py:992]: Use similar naming scheme to the device side one
https://github.com/ROCm/iris/blob/9ff704801e2bcfa5d6fef29d2ae7734fe876a388/iris/mem/triton/ops.py#L281-L299

File: iris/host/iris.py [hunk 10]
```diff
@@ -918,6 +950,306 @@ def get_heap_bases(self):
...
+            # Simple copy
+            sdma_ep.put(src_rank, dst_rank, channel, src_ptr, dst_ptr, size)
+
+        if not async_op:
+            sdma_ep.quiet(src_rank, dst_rank, channel)
+
+    def put_tile(
+        self,
+        tile,
+        dst_rank: int,
```
Comment (mawad-amd) [inline iris/host/iris.py:1084]: same comment on dst_*

File: iris/mem/triton/ops.py [hunk 11]
```diff
@@ -287,23 +289,36 @@ def put(
...
 
+    Supports both 1D (flat/linear) and 2D (tiled) copies:
+    - 1D copies: Used for 1D pointer blocks, uses linear SDMA packets
+    - 2D copies: Used for 2D pointer blocks, uses sub-window SDMA packets for better performance
     The load is **always local** (reading from the current rank's own ``from_ptr``), while the
     store is **remote** when ``from_rank != to_rank`` (writing to a peer GPU).
 
     Args:
         from_ptr (triton.PointerType, or block of dtype=triton.PointerType): Pointer in the current rank's local memory from which to read data.
-        to_ptr (triton.PointerType, or block of dtype=triton.PointerType): Pointer in the current rank's address space that will be translated to the `to_rank`'s address space. Must be the current rank where the pointer is local.
```
Comment (mawad-amd) [inline iris/mem/triton/ops.py:302]: Why was this removed?

File: iris/mem/triton/ops.py [hunk 12]
```diff
@@ -287,23 +289,36 @@ def put(
     load_cache_modifier=None,
     store_cache_modifier=None,
     hint: tl.constexpr = None,
+    copy_engine_ctx: tl.tensor = None,
+    src_row_stride: tl.constexpr = 0,
+    dst_row_stride: tl.constexpr = 0,
+    USE_COPY_ENGINE: tl.constexpr = False,
```
Comment (mawad-amd) [inline iris/mem/triton/ops.py:295]: let's just use lower case for all new arguments/constexpr

File: iris/mem/triton/ops.py [hunk 13]
```diff
@@ -319,27 +334,186 @@ def put(
...
+    val,
+    from_rank,
+    to_rank,
+    heap_bases,
+    mask=None,
+    sem=None,
+    scope=None,
+    hint: tl.constexpr = None,
+    copy_engine_ctx=None,
+    USE_COPY_ENGINE: tl.constexpr = False,
```
Comment (mawad-amd) [inline iris/mem/triton/ops.py:516]: case

File: iris/mem/triton/ops.py [hunk 14]
```diff
@@ -416,7 +628,19 @@ def atomic_sub(
...
+    cmp,
+    val,
+    from_rank,
+    to_rank,
+    heap_bases,
+    sem=None,
+    scope=None,
+    hint: tl.constexpr = None,
+    copy_engine_ctx=None,
+    USE_COPY_ENGINE: tl.constexpr = False,
```
Comment (mawad-amd) [inline iris/mem/triton/ops.py:642]: case

File: pyproject.toml [hunk 15]
```diff
@@ -22,7 +22,7 @@ dependencies = [
     "requests",
     "ruff",
     "tritonblas @ git+https://github.com/ROCm/tritonBLAS.git@muhaawad/iris",
-
+    "rocm-xio @ git+https://github.com/ROCm/rocm-xio.git",
```
Comment (mawad-amd) [inline pyproject.toml:25]: This will likely break the simulation backend but let's merge it and see.

**Review by mawad-amd — APPROVED** [review]
Some minor comments. Thanks a lot for this David!

Please fix/resolve conflict, and just make sure CI passes and merge.
